File: Android-Client/app/src/main/java/com/grabtube/android/domain/model/Schedule.kt
package com.grabtube.android.domain.model

import java.time.Instant

/**
 * Types of schedules supported
 */
enum class ScheduleType {
    ONE_TIME,      // Execute once at specific date/time
    RECURRING,     // Execute repeatedly (daily, weekly, monthly)
    PERIODIC,      // Execute every X time units
    COLLECTION     // For downloading video collections periodically
}

/**
 * Recurrence patterns for recurring schedules
 */
enum class RecurrencePattern {
    DAILY,
    WEEKLY,
    MONTHLY,
    YEARLY
}

/**
 * Days of the week for weekly recurrence
 */
enum class WeekDay {
    MONDAY,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY,
    SUNDAY
}

/**
 * Time units for periodic schedules
 */
enum class TimeUnit {
    MINUTES,
    HOURS,
    DAYS,
    WEEKS,
    MONTHS
}

/**
 * Domain model representing a scheduling configuration
 */
data class Schedule(
    val id: String,
    val name: String,
    val description: String = "",
    val type: ScheduleType,
    val startDate: Instant? = null,        // For one-time schedules
    val startTime: Instant? = null,        // Time of day to execute
    val recurrencePattern: RecurrencePattern? = null,
    val weekDays: List<WeekDay>? = null,   // For weekly recurrence
    val dayOfMonth: Int? = null,           // For monthly recurrence (1-31)
    val interval: Int? = null,             // For periodic schedules (every X units)
    val timeUnit: TimeUnit? = null,        // Unit for periodic schedules
    val endDate: Instant? = null,          // When to stop recurring
    val maxExecutions: Int? = null,        // Maximum number of executions
    val isActive: Boolean = true,
    val createdAt: Instant,
    val lastExecutedAt: Instant? = null,
    val nextExecutionAt: Instant? = null,
    val executionCount: Int = 0,
    val metadata: Map<String, Any>? = null // Additional configuration
) {
    /**
     * Calculate next execution time
     */
    fun calculateNextExecution(from: Instant = Instant.now()): Instant? {
        return when (type) {
            ScheduleType.ONE_TIME -> calculateOneTimeNextExecution(from)
            ScheduleType.RECURRING -> calculateRecurringNextExecution(from)
            ScheduleType.PERIODIC -> calculatePeriodicNextExecution(from)
            ScheduleType.COLLECTION -> calculatePeriodicNextExecution(from)
        }
    }

    private fun calculateOneTimeNextExecution(from: Instant): Instant? {
        return if (startDate != null && startTime != null) {
            // Combine date and time
            val dateTime = startDate.atZone(java.time.ZoneId.systemDefault())
                .withHour(startTime.atZone(java.time.ZoneId.systemDefault()).hour)
                .withMinute(startTime.atZone(java.time.ZoneId.systemDefault()).minute)
                .withSecond(0)
                .withNano(0)
                .toInstant()

            if (dateTime.isAfter(from)) dateTime else null
        } else null
    }

    private fun calculateRecurringNextExecution(from: Instant): Instant? {
        if (recurrencePattern == null || startTime == null) return null

        return when (recurrencePattern!!) {
            RecurrencePattern.DAILY -> calculateDailyNextExecution(from)
            RecurrencePattern.WEEKLY -> calculateWeeklyNextExecution(from)
            RecurrencePattern.MONTHLY -> calculateMonthlyNextExecution(from)
            RecurrencePattern.YEARLY -> calculateYearlyNextExecution(from)
        }
    }

    private fun calculateDailyNextExecution(from: Instant): Instant {
        val fromZdt = from.atZone(java.time.ZoneId.systemDefault())
        val timeZdt = startTime!!.atZone(java.time.ZoneId.systemDefault())

        val today = fromZdt.toLocalDate()
        val executionTime = today.atTime(timeZdt.toLocalTime())

        return if (executionTime.isAfter(fromZdt)) {
            executionTime.toInstant()
        } else {
            // Tomorrow
            val tomorrow = today.plusDays(1)
            tomorrow.atTime(timeZdt.toLocalTime()).toInstant()
        }
    }

    private fun calculateWeeklyNextExecution(from: Instant): Instant? {
        if (weekDays.isNullOrEmpty()) return null

        val fromZdt = from.atZone(java.time.ZoneId.systemDefault())
        val timeZdt = startTime!!.atZone(java.time.ZoneId.systemDefault())

        val currentDayOfWeek = fromZdt.dayOfWeek
        val sortedWeekDays = weekDays.sortedBy { it.ordinal }

        // Find next weekday in the list
        val nextWeekDay = sortedWeekDays.firstOrNull { it.ordinal >= currentDayOfWeek.ordinal }
            ?: sortedWeekDays.first()

        val daysToAdd = nextWeekDay.ordinal - currentDayOfWeek.ordinal
        val targetDate = fromZdt.toLocalDate().plusDays(
            if (daysToAdd < 0) daysToAdd + 7L else daysToAdd.toLong()
        )

        val executionTime = targetDate.atTime(timeZdt.toLocalTime())

        // If the execution time has passed today, move to next occurrence
        return if (executionTime.isBefore(fromZdt) ||
                   (executionTime.isEqual(fromZdt) && nextWeekDay == WeekDay.values()[currentDayOfWeek.ordinal])) {
            val nextTargetDate = targetDate.plusDays(7)
            nextTargetDate.atTime(timeZdt.toLocalTime()).toInstant()
        } else {
            executionTime.toInstant()
        }
    }

    private fun calculateMonthlyNextExecution(from: Instant): Instant? {
        val dayOfMonth = dayOfMonth ?: return null
        val fromZdt = from.atZone(java.time.ZoneId.systemDefault())
        val timeZdt = startTime!!.atZone(java.time.ZoneId.systemDefault())

        val currentMonth = fromZdt.toLocalDate()
        val targetDate = try {
            currentMonth.withDayOfMonth(dayOfMonth)
        } catch (e: Exception) {
            // Invalid day for this month, use last day
            currentMonth.withDayOfMonth(currentMonth.lengthOfMonth())
        }

        val executionTime = targetDate.atTime(timeZdt.toLocalTime())

        return if (executionTime.isAfter(fromZdt)) {
            executionTime.toInstant()
        } else {
            // Next month
            val nextMonth = currentMonth.plusMonths(1)
            val nextTargetDate = try {
                nextMonth.withDayOfMonth(dayOfMonth)
            } catch (e: Exception) {
                nextMonth.withDayOfMonth(nextMonth.lengthOfMonth())
            }
            nextTargetDate.atTime(timeZdt.toLocalTime()).toInstant()
        }
    }

    private fun calculateYearlyNextExecution(from: Instant): Instant {
        val fromZdt = from.atZone(java.time.ZoneId.systemDefault())
        val timeZdt = startTime!!.atZone(java.time.ZoneId.systemDefault())

        val currentYear = fromZdt.year
        val targetDate = java.time.LocalDate.of(currentYear, timeZdt.month, timeZdt.dayOfMonth)
        val executionTime = targetDate.atTime(timeZdt.toLocalTime())

        return if (executionTime.isAfter(fromZdt)) {
            executionTime.toInstant()
        } else {
            // Next year
            val nextYear = currentYear + 1
            val nextTargetDate = java.time.LocalDate.of(nextYear, timeZdt.month, timeZdt.dayOfMonth)
            nextTargetDate.atTime(timeZdt.toLocalTime()).toInstant()
        }
    }

    private fun calculatePeriodicNextExecution(from: Instant): Instant? {
        val interval = interval ?: return null
        val timeUnit = timeUnit ?: return null

        val duration = when (timeUnit) {
            TimeUnit.MINUTES -> java.time.Duration.ofMinutes(interval.toLong())
            TimeUnit.HOURS -> java.time.Duration.ofHours(interval.toLong())
            TimeUnit.DAYS -> java.time.Duration.ofDays(interval.toLong())
            TimeUnit.WEEKS -> java.time.Duration.ofDays(interval.toLong() * 7)
            TimeUnit.MONTHS -> java.time.Duration.ofDays(interval.toLong() * 30) // Approximation
        }

        val lastExecution = lastExecutedAt ?: createdAt
        val nextExecution = lastExecution.plus(duration)

        return if (nextExecution.isAfter(from)) nextExecution else from.plus(duration)
    }

    /**
     * Check if schedule should execute now
     */
    fun shouldExecuteNow(currentTime: Instant = Instant.now()): Boolean {
        val nextExecution = calculateNextExecution(currentTime) ?: return false

        // Allow 1 minute tolerance for execution
        val tolerance = java.time.Duration.ofMinutes(1)
        return java.time.Duration.between(currentTime, nextExecution).abs() <= tolerance
    }

    /**
     * Check if schedule is expired
     */
    fun isExpired(currentTime: Instant = Instant.now()): Boolean {
        if (endDate != null && endDate!!.isBefore(currentTime)) return true
        if (maxExecutions != null && executionCount >= maxExecutions!!) return true

        return false
    }

    companion object {
        /**
         * Create a one-time schedule
         */
        fun oneTime(
            id: String,
            name: String,
            executeAt: Instant,
            description: String = "",
            metadata: Map<String, Any>? = null
        ) = Schedule(
            id = id,
            name = name,
            description = description,
            type = ScheduleType.ONE_TIME,
            startDate = executeAt.truncatedTo(java.time.temporal.ChronoUnit.DAYS),
            startTime = executeAt,
            isActive = true,
            createdAt = Instant.now(),
            metadata = metadata
        )

        /**
         * Create a daily recurring schedule
         */
        fun daily(
            id: String,
            name: String,
            startTime: Instant,
            endDate: Instant? = null,
            maxExecutions: Int? = null,
            description: String = "",
            metadata: Map<String, Any>? = null
        ) = Schedule(
            id = id,
            name = name,
            description = description,
            type = ScheduleType.RECURRING,
            recurrencePattern = RecurrencePattern.DAILY,
            startTime = startTime,
            endDate = endDate,
            maxExecutions = maxExecutions,
            isActive = true,
            createdAt = Instant.now(),
            metadata = metadata
        )

        /**
         * Create a weekly recurring schedule
         */
        fun weekly(
            id: String,
            name: String,
            startTime: Instant,
            weekDays: List<WeekDay>,
            endDate: Instant? = null,
            maxExecutions: Int? = null,
            description: String = "",
            metadata: Map<String, Any>? = null
        ) = Schedule(
            id = id,
            name = name,
            description = description,
            type = ScheduleType.RECURRING,
            recurrencePattern = RecurrencePattern.WEEKLY,
            startTime = startTime,
            weekDays = weekDays,
            endDate = endDate,
            maxExecutions = maxExecutions,
            isActive = true,
            createdAt = Instant.now(),
            metadata = metadata
        )

        /**
         * Create a periodic schedule
         */
        fun periodic(
            id: String,
            name: String,
            interval: Int,
            timeUnit: TimeUnit,
            startDate: Instant? = null,
            endDate: Instant? = null,
            maxExecutions: Int? = null,
            description: String = "",
            metadata: Map<String, Any>? = null
        ) = Schedule(
            id = id,
            name = name,
            description = description,
            type = ScheduleType.PERIODIC,
            interval = interval,
            timeUnit = timeUnit,
            startDate = startDate,
            endDate = endDate,
            maxExecutions = maxExecutions,
            isActive = true,
            createdAt = Instant.now(),
            metadata = metadata
        )

        /**
         * Create a collection schedule for downloading video collections
         */
        fun collection(
            id: String,
            name: String,
            interval: Int,
            timeUnit: TimeUnit,
            collectionUrl: String,
            startDate: Instant? = null,
            endDate: Instant? = null,
            maxExecutions: Int? = null,
            description: String = "",
            metadata: Map<String, Any>? = null
        ) = Schedule(
            id = id,
            name = name,
            description = description,
            type = ScheduleType.COLLECTION,
            interval = interval,
            timeUnit = timeUnit,
            startDate = startDate,
            endDate = endDate,
            maxExecutions = maxExecutions,
            isActive = true,
            createdAt = Instant.now(),
            metadata = (metadata ?: emptyMap()) + ("collectionUrl" to collectionUrl)
        )
    }
}

File: Flutter-Client/lib/domain/entities/schedule.dart
import 'package:equatable/equatable.dart';

/// Schedule type enum
enum ScheduleType {
  oneTime,
  recurring,
  periodic,
  collection,
}

/// Recurrence pattern enum for recurring schedules
enum RecurrencePattern {
  daily,
  weekly,
  monthly,
}

/// Week day enum for weekly schedules
enum WeekDay {
  monday,
  tuesday,
  wednesday,
  thursday,
  friday,
  saturday,
  sunday,
}

/// Time unit enum for periodic schedules
enum TimeUnit {
  minutes,
  hours,
  days,
}

/// Schedule entity representing a scheduled download task
class Schedule extends Equatable {
  const Schedule({
    required this.id,
    required this.name,
    required this.type,
    required this.createdAt,
    this.description,
    this.startDate,
    this.startTime,
    this.recurrencePattern,
    this.weekDays,
    this.interval,
    this.timeUnit,
    this.dayOfMonth,
    this.isActive = true,
    this.lastExecutedAt,
    this.metadata,
  });

  /// Unique schedule identifier
  final String id;

  /// Schedule name
  final String name;

  /// Schedule description (optional)
  final String? description;

  /// Type of schedule (one-time, recurring, periodic, collection)
  final ScheduleType type;

  /// Start date for the schedule (optional, for recurring schedules)
  final DateTime? startDate;

  /// Start time for execution
  final DateTime? startTime;

  /// Recurrence pattern for recurring schedules
  final RecurrencePattern? recurrencePattern;

  /// Days of the week for weekly recurring schedules
  final List<WeekDay>? weekDays;

  /// Interval for periodic schedules
  final int? interval;

  /// Time unit for periodic schedules
  final TimeUnit? timeUnit;

  /// Day of month for monthly recurring schedules (1-31)
  final int? dayOfMonth;

  /// Whether the schedule is active
  final bool isActive;

  /// When the schedule was created
  final DateTime createdAt;

  /// When the schedule was last executed
  final DateTime? lastExecutedAt;

  /// Additional metadata for the schedule
  final Map<String, dynamic>? metadata;

  /// Factory constructor for one-time schedule
  factory Schedule.oneTime({
    required String id,
    required String name,
    required DateTime executeAt,
    String? description,
    Map<String, dynamic>? metadata,
  }) {
    return Schedule(
      id: id,
      name: name,
      description: description,
      type: ScheduleType.oneTime,
      startDate: DateTime(executeAt.year, executeAt.month, executeAt.day),
      startTime: executeAt,
      createdAt: DateTime.now(),
      metadata: metadata,
    );
  }

  /// Factory constructor for daily recurring schedule
  factory Schedule.daily({
    required String id,
    required String name,
    required DateTime startTime,
    String? description,
    DateTime? startDate,
    Map<String, dynamic>? metadata,
  }) {
    return Schedule(
      id: id,
      name: name,
      description: description,
      type: ScheduleType.recurring,
      recurrencePattern: RecurrencePattern.daily,
      startDate: startDate,
      startTime: startTime,
      createdAt: DateTime.now(),
      metadata: metadata,
    );
  }

  /// Factory constructor for weekly recurring schedule
  factory Schedule.weekly({
    required String id,
    required String name,
    required DateTime startTime,
    required List<WeekDay> weekDays,
    String? description,
    DateTime? startDate,
    Map<String, dynamic>? metadata,
  }) {
    return Schedule(
      id: id,
      name: name,
      description: description,
      type: ScheduleType.recurring,
      recurrencePattern: RecurrencePattern.weekly,
      weekDays: weekDays,
      startDate: startDate,
      startTime: startTime,
      createdAt: DateTime.now(),
      metadata: metadata,
    );
  }

  /// Factory constructor for periodic schedule
  factory Schedule.periodic({
    required String id,
    required String name,
    required int interval,
    required TimeUnit timeUnit,
    String? description,
    Map<String, dynamic>? metadata,
  }) {
    return Schedule(
      id: id,
      name: name,
      description: description,
      type: ScheduleType.periodic,
      interval: interval,
      timeUnit: timeUnit,
      createdAt: DateTime.now(),
      metadata: metadata,
    );
  }

  /// Factory constructor for collection schedule (e.g., playlist monitoring)
  factory Schedule.collection({
    required String id,
    required String name,
    required int interval,
    required TimeUnit timeUnit,
    required String collectionUrl,
    String? description,
  }) {
    return Schedule(
      id: id,
      name: name,
      description: description,
      type: ScheduleType.collection,
      interval: interval,
      timeUnit: timeUnit,
      createdAt: DateTime.now(),
      metadata: {'collectionUrl': collectionUrl},
    );
  }

  /// Calculate the next execution time for this schedule
  DateTime? calculateNextExecution() {
    if (!isActive) return null;

    final now = DateTime.now();
    final baseTime = lastExecutedAt ?? createdAt;

    switch (type) {
      case ScheduleType.oneTime:
        if (startTime == null) return null;
        return startTime!.isAfter(now) ? startTime : null;

      case ScheduleType.recurring:
        return _calculateRecurringExecution(now);

      case ScheduleType.periodic:
      case ScheduleType.collection:
        return _calculatePeriodicExecution(baseTime);
    }
  }

  /// Calculate next execution for recurring schedules
  DateTime? _calculateRecurringExecution(DateTime now) {
    if (startTime == null) return null;

    switch (recurrencePattern) {
      case RecurrencePattern.daily:
        return _calculateDailyExecution(now);

      case RecurrencePattern.weekly:
        return _calculateWeeklyExecution(now);

      case RecurrencePattern.monthly:
        return _calculateMonthlyExecution(now);

      default:
        return null;
    }
  }

  /// Calculate next daily execution
  DateTime _calculateDailyExecution(DateTime now) {
    final targetTime = DateTime(
      now.year,
      now.month,
      now.day,
      startTime!.hour,
      startTime!.minute,
    );

    if (targetTime.isAfter(now)) {
      return targetTime;
    } else {
      return targetTime.add(const Duration(days: 1));
    }
  }

  /// Calculate next weekly execution
  DateTime _calculateWeeklyExecution(DateTime now) {
    if (weekDays == null || weekDays!.isEmpty) {
      return _calculateDailyExecution(now);
    }

    final currentWeekDay = now.weekday;
    final targetWeekDays = weekDays!.map(_weekDayToInt).toList()..sort();

    for (final targetDay in targetWeekDays) {
      int daysUntilTarget = targetDay - currentWeekDay;
      if (daysUntilTarget < 0) daysUntilTarget += 7;

      final targetDate = now.add(Duration(days: daysUntilTarget));
      final targetTime = DateTime(
        targetDate.year,
        targetDate.month,
        targetDate.day,
        startTime!.hour,
        startTime!.minute,
      );

      if (targetTime.isAfter(now)) {
        return targetTime;
      }
    }

    // If no valid time today, find next week's first occurrence
    final firstDay = targetWeekDays.first;
    int daysUntilNext = firstDay - currentWeekDay;
    if (daysUntilNext <= 0) daysUntilNext += 7;

    final nextDate = now.add(Duration(days: daysUntilNext));
    return DateTime(
      nextDate.year,
      nextDate.month,
      nextDate.day,
      startTime!.hour,
      startTime!.minute,
    );
  }

  /// Calculate next monthly execution
  DateTime _calculateMonthlyExecution(DateTime now) {
    if (dayOfMonth == null) return _calculateDailyExecution(now);

    final targetDate = DateTime(
      now.year,
      now.month,
      dayOfMonth!,
      startTime!.hour,
      startTime!.minute,
    );

    if (targetDate.isAfter(now)) {
      return targetDate;
    } else {
      // Next month
      final nextMonth = now.month == 12 ? 1 : now.month + 1;
      final nextYear = now.month == 12 ? now.year + 1 : now.year;
      return DateTime(
        nextYear,
        nextMonth,
        dayOfMonth!,
        startTime!.hour,
        startTime!.minute,
      );
    }
  }

  /// Calculate next periodic execution
  DateTime _calculatePeriodicExecution(DateTime baseTime) {
    if (interval == null || timeUnit == null) return DateTime.now();

    final duration = _getDuration(interval!, timeUnit!);
    return baseTime.add(duration);
  }

  /// Convert time unit and interval to Duration
  Duration _getDuration(int interval, TimeUnit unit) {
    switch (unit) {
      case TimeUnit.minutes:
        return Duration(minutes: interval);
      case TimeUnit.hours:
        return Duration(hours: interval);
      case TimeUnit.days:
        return Duration(days: interval);
    }
  }

  /// Convert WeekDay to int (1 = Monday, 7 = Sunday)
  int _weekDayToInt(WeekDay day) {
    switch (day) {
      case WeekDay.monday:
        return 1;
      case WeekDay.tuesday:
        return 2;
      case WeekDay.wednesday:
        return 3;
      case WeekDay.thursday:
        return 4;
      case WeekDay.friday:
        return 5;
      case WeekDay.saturday:
        return 6;
      case WeekDay.sunday:
        return 7;
    }
  }

  /// Check if schedule should execute now (within 1 minute window)
  bool shouldExecuteNow() {
    final nextExecution = calculateNextExecution();
    if (nextExecution == null) return false;

    final now = DateTime.now();
    final difference = nextExecution.difference(now);

    return difference.inSeconds.abs() <= 60; // Within 1 minute
  }

  /// Check if one-time schedule is expired
  bool isExpired() {
    if (type != ScheduleType.oneTime) return false;
    if (startTime == null) return true;

    return startTime!.isBefore(DateTime.now());
  }

  /// Create a copy with updated fields
  Schedule copyWith({
    String? id,
    String? name,
    String? description,
    ScheduleType? type,
    DateTime? startDate,
    DateTime? startTime,
    RecurrencePattern? recurrencePattern,
    List<WeekDay>? weekDays,
    int? interval,
    TimeUnit? timeUnit,
    int? dayOfMonth,
    bool? isActive,
    DateTime? createdAt,
    DateTime? lastExecutedAt,
    Map<String, dynamic>? metadata,
  }) {
    return Schedule(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      type: type ?? this.type,
      startDate: startDate ?? this.startDate,
      startTime: startTime ?? this.startTime,
      recurrencePattern: recurrencePattern ?? this.recurrencePattern,
      weekDays: weekDays ?? this.weekDays,
      interval: interval ?? this.interval,
      timeUnit: timeUnit ?? this.timeUnit,
      dayOfMonth: dayOfMonth ?? this.dayOfMonth,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      lastExecutedAt: lastExecutedAt ?? this.lastExecutedAt,
      metadata: metadata ?? this.metadata,
    );
  }

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        type,
        startDate,
        startTime,
        recurrencePattern,
        weekDays,
        interval,
        timeUnit,
        dayOfMonth,
        isActive,
        createdAt,
        lastExecutedAt,
        metadata,
      ];
}
